test-common/utils.c

```diff
@@ -1846,6 +1883,266 @@ get_update_server_dir (GFile *client_root)
+  g_autoptr(GFile) flatpak_autoinstall_override_dir = get_flatpak_autoinstall_override_dir (updater_dir);
+  g_autoptr(GFile) sysroot = get_sysroot_for_client (client_root);
+  g_autofree gchar *sysroot_path = g_file_get_path (sysroot);
+  g_autofree gchar *deployment_id = g_strdup_printf("%s.0", deployment_csum);
+  g_autofree gchar *deployment_datadir = g_build_filename (sysroot_path,
+                                                           "ostree",
+                                                           "deploy",
+                                                           remote,
+                                                           "deploy",
+                                                           deployment_id,
+                                                           "usr",
+                                                           "share",
+                                                           NULL);
+  g_autoptr(GFile) datadir = g_file_new_for_path (deployment_datadir);
+
+  CmdArg args[] = {
+    { NULL, eos_flatpak_installer_binary },
+    { NULL, NULL }
+  };
+  CmdEnvVar envv[] =
+    {
+      { "EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR", NULL, flatpak_installation_dir },
+      { "EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR", NULL, flatpak_upgrade_state_dir },
+      { "EOS_UPDATER_TEST_UPDATER_FLATPAK_AUTOINSTALL_OVERRIDE_DIRS", NULL, flatpak_autoinstall_override_dir },
+      { "EOS_UPDATER_TEST_OSTREE_DATADIR", NULL, datadir },
+      { "EOS_UPDATER_TEST_OVERRIDE_ARCHITECTURE", "arch", NULL },
+      { NULL, NULL, NULL }
+    };
+
+  g_auto(GStrv) argv = build_cmd_args (args);
+  g_auto(GStrv) envp = build_cmd_env (envv);
+
+  if (!test_spawn ((const gchar * const *) argv,
+                   (const gchar * const *) envp,
+                   &cmd,
+                   error))
+    return FALSE;
+
+  return cmd_result_ensure_ok (&cmd, error);
+}
+
+static GStrv
+g_hash_set_to_strv (GHashTable *string_hash_set)
+{
+  GStrv strv = g_new0 (gchar *, g_hash_table_size (string_hash_set) + 1);
+  GList *iter = g_hash_table_get_keys (string_hash_set);
+  gsize i = 0;
+
+  for (; iter; iter = iter->next, ++i)
+    strv[i] = g_strdup (iter->data);
+
+  return strv;
+}
+
+GStrv
+eos_test_get_installed_flatpaks (GFile   *updater_dir,
+                                 GError **error)
+{
+  CmdResult cmd = CMD_RESULT_CLEARED;
+  g_auto(GStrv) installed_flatpaks_lines = NULL;
+  GStrv installed_flatpaks_lines_iter = NULL;
+  g_autoptr(GHashTable) installed_flatpaks_set = g_hash_table_new_full (g_str_hash,
+                                                                        g_str_equal,
+                                                                        g_free,
+                                                                        NULL);
+  g_autoptr(GRegex) flatpak_id_regex = g_regex_new ("(.*?)/.*?/.*?", 0, 0, error);
+
+  if (!flatpak_id_regex)
```

Comment:
Nitpick: `flatpak_id_regex == NULL`.

Reply:
Actually, it might simplify things (and make them more robust) to use `flatpak_ref_parse()` rather than a regex. Split each line by whitespace, then use `flatpak_ref_parse()` on the first field. I’m probably going to push this commit as-is (with the nitpicks fixed), but if you could do a follow-up commit to change it to use `flatpak_ref_parse()` that would be great.

Reply:
Hmm, so the reason why I didn't use that here is that in the "acceptance tests" (e.g., test-common) we have not been linking to libflatpak directly and instead been using the CLI.

I'd be fine using the library, but only if that's what we want to do.

Reply:
Aah, that makes a lot of sense. I’ll add a comment to that effect above the regex. Thanks.